pkg/notify/notifier/discord/discord.go

```diff
@@ -0,0 +1,196 @@
+package discord
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/kubesphere/notification-manager/pkg/async"
+	"github.com/kubesphere/notification-manager/pkg/constants"
+	"github.com/kubesphere/notification-manager/pkg/controller"
+	"github.com/kubesphere/notification-manager/pkg/internal"
+	"github.com/kubesphere/notification-manager/pkg/internal/discord"
+	"github.com/kubesphere/notification-manager/pkg/notify/notifier"
+	"github.com/kubesphere/notification-manager/pkg/template"
+	"github.com/kubesphere/notification-manager/pkg/utils"
+	"net/http"
+	"strings"
+	"time"
+)
+
+const (
+	DefaultSendTimeout  = time.Second * 3
+	DefaultTextTemplate = `{{ template "nm.default.text" . }}`
+	MaxContentLength    = 2000
+	EmbedLimit          = 4096
+)
+
+type Notifier struct {
+	notifierCtl *controller.Controller
+	receiver    *discord.Receiver
+	timeout     time.Duration
+	logger      log.Logger
+	tmpl        *template.Template
+}
+type Message struct {
+	Content string   `json:"content"`
+	Embeds  []Embeds `json:"embeds,omitempty"`
+}
+
+type Embeds struct {
+	Description string `json:"description,omitempty"`
+}
+
+func NewDiscordNotifier(logger log.Logger, receiver internal.Receiver, notifierCtl *controller.Controller) (notifier.Notifier, error) {
+
+	n := &Notifier{
+		notifierCtl: notifierCtl,
+		logger:      logger,
+		timeout:     DefaultSendTimeout,
+	}
+
+	opts := notifierCtl.ReceiverOpts
+	tmplType := constants.Text
+	tmplName := ""
+	if opts != nil && opts.Global != nil && !utils.StringIsNil(opts.Global.Template) {
+		tmplName = opts.Global.Template
+	}
+
+	if opts != nil && opts.Discord != nil {
+		if !utils.StringIsNil(opts.Discord.Template) {
+			tmplName = opts.Discord.Template
+		}
+
+		if opts.Discord.NotificationTimeout != nil {
+			n.timeout = time.Second * time.Duration(*opts.Discord.NotificationTimeout)
+		}
+	}
+
+	n.receiver = receiver.(*discord.Receiver)
+	if utils.StringIsNil(n.receiver.TmplType) {
+		n.receiver.TmplType = tmplType
+	}
+
+	if utils.StringIsNil(n.receiver.TmplName) {
+		if tmplName != "" {
```

Comment:
Can the `content` and `embed` messages use the same template?

Reply:
yes,they only differ in the display.
content:
<img width="869" alt="image" src="https://user-images.githubusercontent.com/35139861/209915389-ba8feec5-7448-418b-847e-6ecf6c738ab6.png">
embed:
<img width="608" alt="image" src="https://user-images.githubusercontent.com/35139861/209915356-d2cfee3b-c506-4717-905c-663098616ea3.png">